dali/kernels/imgproc/color_manipulation/hsv_cpu.h

```diff
@@ -0,0 +1,139 @@
+ *  image.y |      +-----+                   |
+ *       |  | roi.y|     |                   |
+ *       v  |      +-----+                   |
+ *          +--------------------------------+
+ */
+template<size_t ndims>
+using Roi = Box<ndims, int>;
+
+
+/**
+ * Defines TensorShape corresponding to provided Roi.
+ * Assumes HWC memory layout
+ *
+ * @tparam ndims_roi Number of dims in Roi
+ * @param roi Region of interest
+ * @param nchannels Number of channels in data
+ * @return Corresponding TensorShape
+ */
+template<size_t ndims_roi>
+TensorShape<ndims_roi + 1> roi_shape(Roi<ndims_roi> roi, size_t nchannels) {
+    assert(all_coords(roi.hi >= roi.lo) && "Cannot create a tensor shape from an invalid Box");
+    TensorShape<ndims_roi + 1> ret;
+    auto e = roi.extent();
+    auto ridx = ndims_roi;
+    ret[ridx--] = nchannels;
+    for (size_t idx = 0; idx < ndims_roi; idx++) {
+        ret[ridx--] = e[idx];
+    }
+    return ret;
+}
+
+}  // namespace hsv
+
+
+template<class OutputType, class InputType>
+class HsvCpu {
+ public:
+    using Roi = hsv::Roi<hsv::ndims - 1>;
+
+
+    KernelRequirements
+    Setup(KernelContext &context, const InTensorCPU<InputType, hsv::ndims> &in, float hue,
+          float saturation, float value, const Roi *roi = nullptr) {
+        DALI_ENFORCE(!roi || all_coords(roi->hi >= roi->lo), "Region of interest is invalid");
+        auto adjusted_roi = AdjustRoi(roi, in.shape);
+        KernelRequirements req;
+        TensorListShape<> out_shape({hsv::roi_shape(adjusted_roi, hsv::nchannels)});
+        req.output_shapes = {std::move(out_shape)};
+        return req;
+    }
+
+
+    void Run(KernelContext &context, const OutTensorCPU<OutputType, hsv::ndims> &out,
+             const InTensorCPU<InputType, hsv::ndims> &in, float hue, float saturation, float value,
+             const Roi *roi = nullptr) {
+        auto adjusted_roi = AdjustRoi(roi, in.shape);
+        auto num_channels = in.shape[2];
+        auto image_width = in.shape[1];
+        auto ptr = out.data;
+
+        ptrdiff_t row_stride = image_width * num_channels;
+        auto *row = in.data + adjusted_roi.lo.y * row_stride;
+        for (int y = adjusted_roi.lo.y; y < adjusted_roi.hi.y; y++) {
+            for (int x = adjusted_roi.lo.x; x < adjusted_roi.hi.x; x++) {
+                *ptr++ = ConvertSat<OutputType>(row[x * num_channels + 0] + hue);
+                *ptr++ = ConvertSat<OutputType>(row[x * num_channels + 1] * saturation);
+                *ptr++ = ConvertSat<OutputType>(row[x * num_channels + 2] * value);
+            }
+            row += row_stride;
+        }
+    }
+
+
+ private:
+    /**
+     * Adjusted Roi is a Roi, which doesn't overflow the image,
+     * that given by TensorShape.
+     *
+     * In case, when no Roi is provided (roi == nullptr),
+     * size of whole image is returned as Roi.
+     *
+     * Assumes HWC layout
+     */
+    Roi AdjustRoi(const Roi *roi, const TensorShape<hsv::ndims> &shape) {
```

Comment:
I'll quote myself here ;)
> Not really, because this function also takes care of the case, where ROI is not provided (then, returned Roi is of the size of the image). So it's not really clipping anything, rather expanding